design/src/rfcs/rfc0024_request_id.md

```diff
@@ -0,0 +1,132 @@
+RFC: RequestID in business logic handlers
+=============
+
+> Status: RFC
+>
+> Applies to: server
+
+For a summarized list of proposed changes, see the [Changes Checklist](#changes-checklist) section.
+
+Terminology
+-----------
+
+- **RequestID**: a service-wide unique identifier
+- **UUID**: a universally unique identifier
+
+RequestID is an element that uniquely identifies a client request. RequestID is used by services to map all logs, events and
+specific data to a single operation. This RFC discusses whether and how smithy-rs can make that value available to customers.
+
+Services use a RequestID to collect logs related to the same request and see its flow through the various operations,
+help clients debug requests by sharing this value and, in some cases, use this value to perform their business logic. RequestID is unique across a service at least within a certain timeframe.
```

Comment:
Should the server-generated request ID be echoed back to the client as a header on the returned response?

Reply:
Yes: this is a feature we will add with IntoParts (for adding headers to responses)

Reply:
I am **not in favor** of doing this. 1/ The other day we discussed the Postel's law "Be conservative in what you send; be liberal in what you accept.". The specs do not require us to send this header. 2/ Up to release 0.51, smithy-rs was a pure http / smithy compliant server, however always sending back a custom header breaks that compliance 3/ How would a service owner disable this feature as they might not want this header to be sent? 4/ If a client wishes to trace their requests on the server, it should be their responsibility to send a proper `x-amz-request-id`. 5/ I think it would be better to have this as a middleware layer which service owners can opt into, and they should be able to specify the header they want sent.

Reply:
Sorry I misunderstood that it will always be sent. Since service owners can decide if they want to opt-in or not, this is perfectly fine to go ahead and do this then.